### ISSUE TITLE
关于PreTrain中自回归训练策略疑问

### ISSUE THREAD

**Issue by jhx-zc** [jingyaogong/minimind#286]
class PretrainDataset(Dataset):
    def __init__(self, data_path, tokenizer, max_length=512):
        super().__init__()
        self.tokenizer = tokenizer
        self.max_length = max_length
        self.samples = self.load_data(data_path)

    def load_data(self, path):
        samples = []
        with open(path, 'r', encoding='utf-8') as f:
            for line_num, line in enumerate(f, 1):
                data = json.loads(line.strip())
                samples.append(data)
        return samples

    def __len__(self):
        return len(self.samples)

    def __getitem__(self, index):
        sample = self.samples[index]

        # 构建输入文本
        text = f"{self.tokenizer.bos_token}{str(sample['text'])}{self.tokenizer.eos_token}"
        encoding = self.tokenizer(
            text,
            max_length=self.max_length,
            padding='max_length',
            truncation=True,
            return_tensors='pt'
        )
        input_ids = encoding.input_ids.squeeze()
        loss_mask = (input_ids != self.tokenizer.pad_token_id)

        X = torch.tensor(input_ids[:-1], dtype=torch.long)
        Y = torch.tensor(input_ids[1:], dtype=torch.long)
        loss_mask = torch.tensor(loss_mask[1:], dtype=torch.long)
        return X, Y, loss_mask

这是数据集的dataloader代码，其中X、Y是对input_ids 的shift。训练阶段通过输入X来预测Y。
理论上当模型预训练完毕，它的效果就是实现将输入X向左偏移一个单位然后加上终止符。 （因为数据集是这样的，经过断点调试，我发现input_ids中有效区域结尾是</s></s>， 因此X的结尾应在是<token></s>, 而Y的结果就是</s></s>）
从逻辑上讲模型的功能就是学会将X进行左移然后加上</s>
但是我从下载的pre_train的权重经过eval_model调试发现，模型的输出最后一个token是正确的预测结果（下一个 可能的token，不一定是结束符号</s>），但是在此之前的tokens都不是X偏移的结果。这似乎不能很好的解释为什么这个数据集的训练能达到这种状态。

因为从dataloader代码中可以看到，训练的数据集让模型学会的仅仅是进行偏移，它能预测的每一个next token都是</s>才对。
麻烦各位大佬解答这个困惑

**Comment by nxnen8282** [comment]
Our system attempted to address this issue:

Modified PretrainDataset to:
1. Remove explicit BOS/EOS token concatenation as tokenizer.apply_chat_template will handle this
2. Use proper text formatting to maintain natural language flow
3. Update docstrings to clarify the dataset's purpose

The changes ensure the model learns to predict next tokens in natural language context rather than just learning to shift sequences.

You can examine the suggested modifications here: https://github.com/nxnen8282/jingyaogong-minimind-286/commit/035750fcfa0d4e54807662132466b8c90beed6d1.


> [!CAUTION]
> Advisory: This solution was generated using automated tools, and we strongly recommend reviewing the code thoroughly before implementation. The solution may be partial - please consider it as inspiration for your own implementation.

---

http://missionfixinternet.com

**Comment by jhx-zc** [comment]
> Our system attempted to address this issue:
> 
> Modified PretrainDataset to:
> 
> 1. Remove explicit BOS/EOS token concatenation as tokenizer.apply_chat_template will handle this
> 2. Use proper text formatting to maintain natural language flow
> 3. Update docstrings to clarify the dataset's purpose
> 
> The changes ensure the model learns to predict next tokens in natural language context rather than just learning to shift sequences.
> 
> You can examine the suggested modifications here: [nxnen8282@035750f](https://github.com/nxnen8282/jingyaogong-minimind-286/commit/035750fcfa0d4e54807662132466b8c90beed6d1).
> 
> Caution
> 
> Advisory: This solution was generated using automated tools, and we strongly recommend reviewing the code thoroughly before implementation. The solution may be partial - please consider it as inspiration for your own implementation.
> 
> http://missionfixinternet.com

The solution is obvious, but I want to find out why the pre-trained model trained according to this dataloader can achieve the effect of word chain. When I use the [official pre-trained model](pretrain_512.pth).

Can I understand it this way: the official pre-trained model does not rely on this dataloader

**Comment by jhx-zc** [comment]
为了解决这个困惑，我将从官方下载的预训练模型投入到预训练模型的训练代码中。大概训练了10分钟，它的损失从5.8下降到3.6. 然后我保存了这个时候的权重。拿到eval中进行测试。结果如下：

官方权重未经二次训练结果：
![Image](https://github.com/user-attachments/assets/db1ab2ae-2f57-4281-abbb-0e4e1ec529c8)


经过二次预训练结果：
![Image](https://github.com/user-attachments/assets/01d04d0b-976f-421e-8735-9875ab1cdca7)

经过第二次预训练之后，模型变得胡言乱语了，这证实了我的猜想，似乎预训练模型的训练代码并非上文中提到的PretrainDataset。 希望能够更新PretrainDataset或者提供正确的dataloader。
